draft-ietf-tls-ticketrequests.md

```diff
@@ -124,32 +146,67 @@ Clients MAY send this extension in ClientHello. It contains the following struct
 
 ~~~
 struct {
-    uint8 count;
-} TicketRequestContents;
+    uint8 new_session_count;
+    uint8 resumption_count;
+} ClientTicketRequest;
 ~~~
 
-count
-: The number of tickets desired by the client.
+new_session_count
+: The number of tickets desired by the client when the server chooses to
+negotiate a fresh session (full handshake).
+
+resumption_count
+: The number of tickets desired by the client when the server is willing to
+resume using the presented ticket.
+
+A client starting a fresh connection SHOULD set new_session_count to the desired
+number of session tickets and resumption_count to 0.
+Once a client's ticket cache is primed, a resumption_count of 1 is a
+good choice that allows the server to replace each ticket with a fresh ticket,
+without over-provisioning the client with excess tickets. However, clients
+which race multiple connections and place a separate ticket in each will
+ultimately end up with just the tickets from a single resumed session.
+In that case, clients can send a resumption_count equal to the number of
+sessions they are attempting in parallel.
+
+When a client presenting a previously obtained ticket finds that the server
+nevertheless negotiates a fresh session, the client SHOULD assume that any
+other tickets associated with the same session as the presented ticket are also
+no longer valid for resumption.  This includes tickets obtained
+during the initial full handshake and all tickets subsequently obtained as
+part of subsequent resumptions.  Requesting more than one ticket in cases when
+servers select a full handshake helps keep the session cache primed.
+
+Servers SHOULD NOT send more tickets than requested for the handshake type
+selected by the server (resumption or full handshake). Moreover, servers
+SHOULD place a limit on the number of tickets they are willing to send, whether
+for full handshakes or resumptions, to save resources.  Therefore, the
+number of NewSessionTicket messages sent will typically be the minimum
+of the server's self-imposed limit and the number requested.
+Servers MAY send additional tickets, up to the same limit, if the tickets
+that are originally sent are somehow invalidated.
+
+A server which supports and uses a client "ticket_request" extension MUST also send
```

Comment:
What's the rationale for this being a normative MUST? I.e., what would happen if the server didn't send the extension, but did provide the number of tickets that the client wanted? That would look like a server that didn't want to use the request, but got lucky. It seems like it could be a SHOULD, since it's hard to enforce?

Reply:
Less options (and therefore consistency) for server implementers, mainly. This makes the behavior clearly, "if you use this extension, send one in response."

Reply:
Right, makes sense. Sounds like a SHOULD, though. MUST feels like something that can result in a protocol violation error, which seems (to me right now) to be hard to enforce.

Reply:
If it's a SHOULD and a server doesn't do it then clients can't determine whether or not the hint was consumed, which is the point of echoing it back. As a MUST, its use is clear.

Reply:
(I don't claim that this is useful for the client to know, as MT points out. Only that there's ambiguity if it's a SHOULD.)

Reply:
I generally like the rule of thumb that a MUST needs to be enforceable; you can point to an implementation that doesn't follow the MUST and identify it. 

As you say, echoing back the extension is a hint that the request was consumed, and clients MUST be able to handle it not being present (in case it wasn't actually used or consumed). Therefore it is impossible to distinguish between a server that just didn't consume the request, and a "bad" server that ignored your MUST flagrantly. I agree that we definitely want servers to send this, but it feels like a SHOULD since it doesn't allow anyone to detect misbehavior or send an error.

Reply:
The reason I would use for making this a "MUST" is that it ensures that clients have a better idea about when new tickets are likely to be replacements.  If the server says 2 and then sends a third ticket, the client might infer that the first ticket it received is no longer good.

Reply:
I think I would prefer a MUST here as well. That way, if a server does not send it we know that it is not compliant to this specification. Also, when implementing the server part of this, engineers don't have to decide whether to send this or not.

Reply:
Eh, okay. I still think that it's clear enough as a SHOULD, but I can live with it.